common/src/main/java/io/netty/util/internal/CleanerJava24.java

```diff
@@ -0,0 +1,166 @@
+                // The 'allocate' method takes a 'long' capacity but we'll be providing an 'int'.
+                // Explicitly cast the 'long' to 'int' so we can use 'invokeExact'.
+                // allocateInt.type() = (Arena,int)MemorySegment
+                MethodHandle allocateInt = MethodHandles.explicitCastArguments(allocate,
+                        methodType(memsegCls, arenaCls, int.class));
+                // Use the 'asByteBuffer' method as a filter, to transform the constructor into a method that takes a
+                // MemorySegment argument instead of a ByteBuffer argument.
+                // ctorArenaMemseg.type() = (Arena,MemorySegment)CleanableDirectBufferImpl
+                MethodHandle ctorArenaMemseg = MethodHandles.explicitCastArguments(
+                        MethodHandles.filterArguments(bufClsCtor, 1, asByteBuffer),
+                        methodType(bufCls, arenaCls, memsegCls));
+                // Use the 'allocateInt' method to transform the last MemorySegment argument of the constructor,
+                // into an (Arena,int) argument pair.
+                // ctorArenaArenaInt.type() = (Arena,Arena,int)CleanableDirectBufferImpl
+                MethodHandle ctorArenaArenaInt = MethodHandles.collectArguments(ctorArenaMemseg, 1, allocateInt);
+                // Our method now takes two Arena arguments, but we actually only want to pass one. Specifically, it's
+                // very important that it's the same arena we use for both allocation and deallocation.
+                // We permute the argument array such that the first Arena argument gets passed to both parameters,
+                // and the second parameter value gets ignored.
+                // ctorArenaNullInt.type() = (Arena,Arena,int)CleanableDirectBufferImpl
+                MethodHandle ctorArenaNullInt = MethodHandles.permuteArguments(ctorArenaArenaInt,
+                        methodType(bufCls, arenaCls, arenaCls, int.class), 0, 0, 2);
+                // With the second Arena parameter value ignored, we can statically bind it to 'null' to effectively
+                // drop it from our parameter list.
+                // ctorArenaInt.type() = (Arena,int)CleanableDirectBufferImpl
+                MethodHandle ctorArenaInt = MethodHandles.insertArguments(ctorArenaNullInt, 1, new Object[]{null});
+                // Now we just need to create our Arena instance. We fold the Arena parameter into the 'ofShared'
+                // static method, so we effectively bind the argument to the result of calling that method.
+                // Since 'ofShared' takes no further parameters, we effectively eliminate the first parameter.
+                // This creates our method handle that takes an 'int' and returns a 'CleanableDirectBufferImpl'.
+                // ctorInt.type() = (int)CleanableDirectBufferImpl
+                method = MethodHandles.foldArguments(ctorArenaInt, ofShared);
```

Comment:
@franz1981 I tried these method handle shenanigans in a quick microbenchmark, and found the performance to be identical to the equivalent Java code.

Just calling it out, as I know you'll be wondering about that :)